src/main/java/it/pagopa/pn/deliverypush/service/impl/WebhookServiceImpl.java

```diff
@@ -267,56 +250,36 @@ public Mono<Void> purgeEvents(String streamId, String eventId, boolean olderThan
-     *
-     * @param lastEventIdCONFIRMED ultimo evento CONFERMATO su cui basarsi per ricavare il timestamp
-     * @param lastEventIdOLDEST ultimo evento LETTO DA DB su cui basarsi per ricavare il timestamp
-     * @return un timestamp sotto forma di stringa
-     */
-    private String computeLastEventIdWithBufferForPurge(String lastEventIdCONFIRMED, String lastEventIdOLDEST){
-        // se readBuffedDelay  è minore di 0, di fatto è disabilitato
-        if (readBufferDelay <= 0)
-            return lastEventIdCONFIRMED;
 
-        if (StringUtils.hasText(lastEventIdCONFIRMED))
-        {
-            if (StringUtils.hasText(lastEventIdOLDEST))
-            {
-                String timestampCONFIRMED = lastEventIdCONFIRMED.split("_")[0];
-                String timestampOLDEST = lastEventIdOLDEST.split("_")[0];
-                Instant instantCONFIRMED = Instant.parse(timestampCONFIRMED);
-                Instant instantOLDEST = Instant.parse(timestampOLDEST);
-                if (instantOLDEST.isBefore(instantCONFIRMED))
+    private Mono<Void> saveEventWithAtomicIncrement(StreamEntity streamEntity, String iun, String newStatus, String timelineEventCategory, String timelineId, Instant timestamp){
+        // recupero un contatore aggiornato
+        return streamEntityDao.updateAndGetAtomicCounter(streamEntity)
+            .flatMap(atomicCounterUpdated -> {
+                if (atomicCounterUpdated < 0)
                 {
-                    return instantOLDEST.minusMillis(readBufferDelay).toString();
+                    log.warn("updateAndGetAtomicCounter counter is -1, skipping saving stream");
+                    return Mono.empty();
                 }
-            }
 
-            return lastEventIdCONFIRMED;
-        }
-        return null;
+                // creo l'evento e lo salvo
+                EventEntity eventEntity = new EventEntity(atomicCounterUpdated, streamEntity.getStreamId());
+                if (!ttl.isZero())
+                    eventEntity.setTtl(LocalDateTime.now().plus(ttl).atZone(ZoneId.systemDefault()).toEpochSecond());
+
+
+                eventEntity.setEventDescription(timestamp.toString() + "_" + timelineId);
+                eventEntity.setTimestamp(timestamp);
+                // Lo iun ci va solo se è stata accettata, quindi escludo gli stati invalidation e refused
+                if (StringUtils.hasText(newStatus)
+                        && NotificationStatusInt.valueOf(newStatus) != NotificationStatusInt.IN_VALIDATION
+                        && NotificationStatusInt.valueOf(newStatus) != NotificationStatusInt.REFUSED)
+                    eventEntity.setIun(iun);
+                eventEntity.setNewStatus(newStatus);
+                eventEntity.setTimelineEventCategory(timelineEventCategory);
+                // il requestId ci va sempre, ed è il base64 dello iun
+                eventEntity.setNotificationRequestId(Base64Utils.encodeToString(iun.getBytes(StandardCharsets.UTF_8)));
```

Comment:
Da capire se tale campo può essere restituito da delivery, perchè applicare in 2 punti la stessa funziona potrebbe creare problemi in futuro (se si decide di cambiare l'algoritmo ecc.)

Reply:
Vero. 
Attualmente il codice di webhook non chiede nulla a pn-delivery.
Crea pure  una issue in modo che in futuro si possa "risolvere" con la modalità più corretta. Nella issue "linka" il punto del codice  di webhookserviceimpl che così è chiaro il da farsi.